JustKnobsConfig: Make this work with install_config_module

c00w opened the pull request:
The goal of this PR is to make JustKnobsConfig usable with install_config_module. in support of this, there are 3 major parts to this PR.

- The first is that this is based upon https://github.com/pytorch/pytorch/pull/138377, which was broken out into a seperate PR, since it's mostly unrelated to JK changes.
- The second is rewrite install_config_modules to use a layered approach. Explicitly keeping user_overrides and default values seperate. This allows easy insertion of JK default values as a layer in between (and in the future other values if desired, i.e. env variables).
- The last one is convert JustKnobsConfig to be a pure dataclass, and do all resolution within the existing install_config_module resolution system.

There are a couple design decisions that should be mentioned.
1) I flipped everything to a single map, storing _ConfigEntries. We could use multiple maps but this seemed simpler.
2) When saving config we convert everything (including JKs) into user overrides when the configuration is loaded back in. We could do more complex pickling, but that seemed unlikely to have a lot of use.

let me know how this looks.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

ezyang commented on torch/testing/_internal/fake_config_module.py:
```diff
@@ -0,0 +1,11 @@
+example_knob_force= JustKnobsConfig(name="false_knob", default=False, override=True)
+_cache_config_ignore_prefix = []
+
+install_config_module(sys.modules[__name__])
```
So are you going to test this module?

c00w replied:
Yes - sorry forgot to add the files, please take another look.

ezyang reviewed (changes requested):
This seems pretty surprising. It's possible that you have made the right choice, but the PR desc is not enough for me to convinced it should be done this way. And there is not enough exercise of the end-to-end flow for me to evaluate (the test case is too simplified for me to understand)

c00w commented:
@ezyang my apologies, I forget to add the 2 files that actually matter here (Namely the changes to _config_module and _util_internal). Let me know what questions you have with those 2 files pushed.

ezyang commented on torch/_utils_internal.py:
```diff
@@ -182,15 +182,17 @@ class JustKnobsConfig:
 
     def __init__(
-        self, *, name: Optional[str] = None, env_name=None, default: bool = True
+        self, *, name: Optional[str] = None, env_name=None, default: bool = True, override: Optional[bool] = None
```
It's not immediately obvious to me what override does, so it needs docs

ezyang commented on torch/utils/_config_module.py:
```diff
@@ -6,15 +6,15 @@
 
 # Types saved/loaded in configs
-CONFIG_TYPES = (int, float, bool, type(None), str, list, set, tuple, dict)
+CONFIG_TYPES = (int, float, bool, type(None), str, list, set, tuple, dict, JustKnobsConfig)
```
This is really suspicious

c00w replied:
No longer present.

ezyang commented on test/test_utils_internal.py:
```diff
@@ -13,9 +14,46 @@
+            self.assertEqual(out, {"example_knob_force": True})
+        with self.subTest("codegen config"):
+            out = fake_config_module.codegen_config()
+            self.assertEqual(out, "torch.testing._internal.fake_config_module.example_knob_force = True")
```
There's no dependence between these tests. There's no reason to use subTest. Just make separate test methods for each of them if you really want to tell the difference

c00w replied:
Tests gone, but in general removed.

c00w committed:
config: add explicit tests for install_config_module behaviour

c00w committed:
config: remove internal pytorch references to shallow_copy_dict

All use cases upon inspection did not rely on the shallow copy behaviour
(i.e. dicts and other reference values can be modified).

Using the deep copy behaviour makes it possible to remove the shallow
copy behaviour, which makes changing internal implementations almost
impossible.

c00w committed:
config: deprecate shallow copy semantics for config.

These shallow copy semantics prevent changing any internals of the
config module implementation.

Additionally, they only worked on reference type objects (i.e. dicts,
lists) stored within the configuration. All usage with the pytorch repro
did not rely on this functionality.

This change requires release notes for pytorch

c00w committed:
config: Fix config hash not changing when config changes

There is a bug, where the hash is not regenerated if you call get_hash()
then change a config, and then call get_hash() again.

This was masked because the config test utilities set _is_dirty
correctly.

c00w committed:
config: Unify internal dictionary fetching to a single function.

This primarily just makes the inconsistency of the various fetch apis
clear. But also means there is only one fetch location.

c00w committed:
config: Update test patcher to use __setattr__

This update the normal test patcher to use normal __setattr__ methods.
I also removed most of the complexity in the test patcher dirty handling
code, since __setattr__ correctly resets the hash when changed.

c00w committed:
config: teach it to understand JustKnobsConfig

This teaches install_config_module (and the underlying code) to
understands JustKnobsConfig.

This config gets stored within the _ConfigEntry class and is evaluated
when __getattr__ is called.

Due to preceeding work, basically everything works correctly here and we
had to update a couple of tests, and modify the getattr behaviour.

Note that we are overloading the _default value meaning slightly, let me
know if people feel strongly that is a bad idea and I can change it

c00w committed:
config: Modify install_config_module to use a layered approach

This modifies the config system, to use a single mapping of config ->
ConfigEntry and to store the default and user values within them.

We could have used multiple dicts (i.e. user_override and default), but
as we add more fields (justknobs in this PR, perhaps testing and env
variables later), it quickly becomes painful.

There are a couple design decisions we could change.
1) All configs we save store the resolved value - not the default and
   user override seperately
2) All configs we load, apply the resolved value as a user override.

This means that certain complexities of default behvaiour and deletion
(as well as JK), will change if you save + load a config.

c00w commented on test/test_utils_internal.py:
```diff
@@ -13,9 +14,46 @@
 class TestJustKnob(TestCase):
+    def test_justknob_config_module(self):
+        with self.subTest("saving+loading works correctly"):
+            fake_config_module.load_config(fake_config_module.save_config())
```
Test removed, but in general done.

c00w commented:
@ezyang  - PR is updated per offline discussion, if you can take another look, that would be awesome :).

ezyang commented:
Use ghstack to stack diffs and have a usable review view here

c00w commented:
Ack
First new PR - https://github.com/pytorch/pytorch/pull/138758
